### ISSUE TITLE
Upload uses anonymous credentials to PUT to artifactory

### ISSUE THREAD

**Issue by tamboril** [conan-io/conan#979]
I've been unable to get a `conan upload` to work with our Artifactory, and I think I figured out why.
Looking at the request logs, I see that the GET and POST requests have my correct user in them, but the PUT transactions come through with user `anonymous`, and that is rejected with a "403 Forbidden!" message.

If I use curl to do the PUT, it actually creates a file.

I might be able to do a pull request for this, but if you get there first, I think it's in rest_client.py, function `upload_files()`, where it does this:
```
 response = uploader.upload(resource_url, files[filename], auth=auth, dedup=dedup,
                                           retry=retry, retry_wait=retry_wait)
```
I think it needs to pass `self.custom_headers` in there, and then that function needs to use those headers when it does `call_with_retry(...headers=None)` or, if `dedup` is set, merge the custom headers into those it passes to `self.requester.put()`

**Comment by memsharded** [comment]
Seems weird, thanks very much for your feedback and checking this. Which version of Artifactory are you using?  I guess latest conan, right?
Also, did you experience any previous error in auth in previous steps? You can try to clear your credentials with ``$ conan user --clean``. We will have a look asap.

**Comment by tamboril** [comment]
Using Artifactory 5.0.0.  The auth works fine for the `GET` and `POST` actions that take place as part of an upload, but the subsequent `PUT` does not have the same headers, in particular `X-Client-Id`.  Clearing my credentials did not help.  I spent several hours on this.  After stepping through the code and looking at the request logs, I came up with a workaround to allow user `anonymous` permissions "Delete/Overwrite" and "Deploy/Cache" to my Artifactory key.

**Comment by memsharded** [comment]
Checked Artifactory 5.0.1, in fact this is covered by tests, so shouldn't be a problem. I'll try to check 5.0.0 too. What is your deployment? Using proxy? If you want, you can write us with private details to info at conan dot io

**Comment by memsharded** [comment]
Everything looks good with 5.0.0 too (conan version 0.19.1, but should work with previous too)

**Comment by memsharded** [comment]
I think where the problem might come from:
- Conan uses 2 steps transfers, first you have a business logic, then the real transfers. This is to decouple logic from the real server heavy load
- For conan.io it is using signed-urls for the second step. So they have different headers, even for auth.
- There is a ``server_capabilities`` function that checks whether it is communicating with one type of remote or other.
- I think some network settings, like the proxy might be producing a misdetection in the type of remote, then that is the reason of the "anonymous" transfers you are seeing, if you can check, the urls probably have a ``signature`` in them right?

We keep investigating, thanks very much for reporting this and your time checking it!

**Comment by memsharded** [comment]
Confirmed, replicated issue with a forwarding proxy, got 403. Trying to find a fix.

**Comment by tamboril** [comment]
Yes.  We use an nginx forwarding proxy.

**Comment by memsharded** [comment]
Yes, same I tried :)
Opened a minor release with the fix, will try to release it asap.

**Comment by memsharded** [comment]
This has been fixed and released in 0.19.2. Please check.
Thanks very much for your help and work on this issue.
